File: java-api/src/main/java/net/kfoundation/java/serialization/ObjectSerializer.java
// --------------------------------------------------------------------------
//   ██╗  ██╗███████╗
//   ██║ ██╔╝██╔════╝   The KFoundation Project (www.kfoundation.net)
//   █████╔╝ █████╗     KFoundation for Scala Library
//   ██╔═██╗ ██╔══╝     Copyright (c) 2020 Mindscape Inc.
//   ██║  ██╗██║        Terms of KnoRBA Free Public License Agreement Apply
//   ╚═╝  ╚═╝╚═╝
// --------------------------------------------------------------------------

package net.kfoundation.java.serialization;

import net.kfoundation.java.UString;



/**
 * Standard framework for an object serializer.
 */
public interface ObjectSerializer {

    /**
     * Writes a property name (property definition sequence) to output.
     */
    ObjectSerializer writePropertyName(UString name);


    /**
     * Writes a string literal to the output.
     */
    ObjectSerializer writeLiteral(UString value);


    /**
     * Writes an integer literal to the output.
     */
    ObjectSerializer writeLiteral(long value);


    /**
     * Writes a possible fractional decimal number to the ouptut.
     */
    ObjectSerializer writeLiteral(double value);


    /**
     * Writes a boolean literal to the output.
     */
    ObjectSerializer writeLiteral(boolean value);


    /**
     * Writes the symbol representing null to the output.
     */
    ObjectSerializer writeNull();


    /**
     * Writes a sequence representing beginning of an object with the given name
     * to the output.
     */
    ObjectSerializer writeObjectBegin(UString name);


    /**
     * Writes a sequence representing the end of an object to the output.
     */
    ObjectSerializer writeObjectEnd();


    /**
     * Writes the sequence representing the beginning of a collection to the
     * output.
     */
    ObjectSerializer writeCollectionBegin();


    /**
     * Writes the sequence representing the end of a collection to the output.
     */
    ObjectSerializer writeCollectionEnd();


    /**
     * Signals the end of stream.
     */
    void writeStreamEnd();

}

File: shared/src/main/scala/net/kfoundation/scala/serialization/ObjectSerializer.scala
// --------------------------------------------------------------------------
//   ██╗  ██╗███████╗
//   ██║ ██╔╝██╔════╝   The KFoundation Project (www.kfoundation.net)
//   █████╔╝ █████╗     KFoundation for Scala Library
//   ██╔═██╗ ██╔══╝     Copyright (c) 2020 Mindscape Inc.
//   ██║  ██╗██║        Terms of KnoRBA Free Public License Agreement Apply
//   ╚═╝  ╚═╝╚═╝
// --------------------------------------------------------------------------

package net.kfoundation.scala.serialization

import net.kfoundation.scala.UString



/**
 * Abstract framework for object serializers. This class is designed with
 * "fluent" API.
 */
abstract class ObjectSerializer {

  /** Writes a property name to the output. */
  def writePropertyName(name: UString): ObjectSerializer


  /** Writes a string literal to the output. */
  def writeLiteral(value: UString): ObjectSerializer


  /** Writes an integer literal to the output. */
  def writeLiteral(value: Long): ObjectSerializer


  /** Writes a possibly fractional decimal literal to the output. */
  def writeLiteral(value: Double): ObjectSerializer


  /** Writes a boolean value to the output. */
  def writeLiteral(value: Boolean): ObjectSerializer


  /** Writes symbol for null to the output. */
  def writeNull(): ObjectSerializer


  /** Writes the opening part of an object. */
  def writeObjectBegin(name: UString): ObjectSerializer


  /** Writes the closing part of an object. */
  def writeObjectEnd(): ObjectSerializer


  /** Writes the opening part of a collection. */
  def writeCollectionBegin(): ObjectSerializer


  /** Writes the closing part of a collection. */
  def writeCollectionEnd(): ObjectSerializer


  /** Ends the current stream (if applicable). */
  def writeStreamEnd(): Unit


  /**
   * Writes a literal of custom type with string fallback. If this serializer
   * does not directly support the given literal, it uses the given fallback
   * function to produce and write its string equivalent.
   */
  def writeLiteralOrString[T](value: T, fallback: () => UString): ObjectSerializer =
    writeLiteral(fallback())


  /**
   * Writes a literal of custom type with integer fallback. If this serializer
   * does not directly support the given literal, it uses the given fallback
   * function to produce and write its integral equivalent
   */
  def writeLiteralOrInteger[T](value: T, fallback: () => Long): ObjectSerializer =
    writeLiteral(fallback())


  /**
   * Writes a literal of custom type with integer fallback. If this serializer
   * does not directly support the given literal, it uses the given fallback
   * function to produce and write its numeric equivalent
   */
  def writeLiteralOrDecimal[T](value: T, fallback: () => Double): ObjectSerializer =
    writeLiteral(fallback())


  /**
   * Writes given value if a writer for it is available. This method helps
   * using this class in fluent form better.
   */
  def writeValue[T](value: T)(implicit writer: ValueWriter[T]): ObjectSerializer = {
    writer.write(this, value)
    this
  }

}